### ISSUE TITLE
ENH: `np.fix`, `np.isposinf`, and `np.isneginf` as ufuncs

### ISSUE THREAD

**Issue by jakirkham** [numpy/numpy#13100]
Am interested in using `np.fix`, `np.isposinf`, and `np.isneginf` as ordinary `ufunc`s. Since some of the work was already done by adding some deprecations in PR ( https://github.com/numpy/numpy/pull/8996 ), which landed in NumPy 1.13.0, am wondering if it would be possible to complete the conversion. Any thoughts on this?

cc @eric-wieser (who originated the aforementioned PR)

**Comment by eric-wieser** [comment]
It's not clear to me that converting every function to a unfunc is a good idea - they're harder to maintain than to the python functions. Can you summarize why you care whether these are ufuncs?

**Comment by jakirkham** [comment]
Did not know that. Thanks for the info.

The use of `__array_wrap__` with `np.fix` is making it difficult to use Dask ( https://github.com/dask/dask/issues/4496 ) with CuPy ( https://github.com/cupy/cupy/issues/589 ). Converting it to a `ufunc` would make this easier for both libraries to support by leveraging `__array_ufunc__`, which they both already support. AIUI this sort of use case was the [motivation]( https://docs.scipy.org/doc/numpy-1.13.0/neps/ufunc-overrides.html ) for the `__array_ufunc__` protocol. Please feel free to correct me if that is inaccurate.

A bit tangential, but why is it difficult to make and maintain ufuncs? Are there ways to make this less difficult?

**Comment by eric-wieser** [comment]
Right now, `isposinf(x, out)` is implemented as `nx.logical_and(nx.isinf(x), ~nx.signbit(x), out)` - you should be able to override `__array_ufunc__` for each item in turn.

Arguably we should replace `~` with `nx.logical_not` there.

**Comment by jakirkham** [comment]
Thanks @eric-wieser. Have confirmed these work fine locally with Dask.

Have re-scoped the issue to `np.fix` alone.

**Comment by mhvk** [comment]
It is not so obvious to me that having very ufunc-like functions in python makes maintaining or using numpy easier...

Certainly, the python implementation of `isposinf` means the function is about 3 times slower than needed. Indeed,
```
a = np.arange(100000.)
%timeit np.isposinf(a)
# 10000 loops, best of 5: 89.7 µs per loop
%timeit np.equal(a, np.inf)
# 10000 loops, best of 5: 25.1 µs per loop
```

Not sure whether it matters that much for `isposinf`, but for something a bit more commonly used like `isfinite` it might be worth it.

**Comment by eric-wieser** [comment]
@jakirkham: `fix` is once again bitten by the annoying behavior of 0d arrays being unpacked to scalars - that's why that `asanyarray` call is there:
```python
    # promote back to an array if flattened
    res = nx.asanyarray(nx.ceil(x, out=out))
    res = nx.floor(x, out=res, where=nx.greater_equal(x, 0))

    # when no out argument is passed and no subclasses are involved, flatten
    # scalars
    if out is None and type(res) is nx.ndarray:
        res = res[()]
    return res
```

Perhaps we need to add the following:
* A `decay=True` argument to `ufunc.__call__`, where passing `False` disables this unpacking
* An `__array_decay_to_scalar__` protocol, which implements the current unpacking
* An `np.core.decay_to_scalar(x)` function, which calls the protocol if present, else returns `x`

Then we could write `fix` as:
```python
def fix(x, out=None, *, decay=True):
    res = nx.ceil(x, out=out, decay=False)
    res = nx.floor(x, out=res, where=nx.greater_equal(x, 0, decay=False), decay=False)
    if out is None:
        res = np.core.decay_to_scalar(res)
    return res
```

**Comment by seberg** [comment]
Just so everyone is on the same page, what happens is that ufuncs and many other functions use the `PyArray_Return` function. That converts 0-D arrays to scalars. Just as everything else, ufuncs internally only use arrays, so would otherwise naturally return arrays always.

Not sure I like the name `decay` but exposing a way to circumvent `PyArray_Return` would make life easier/things more consistent in a few places. Does dask, etc. actually "decay" to scalars? My first gut feeling is that creating a protocol for something that we would actually like to get rid off if it was possible is strange.

**Comment by mhvk** [comment]
And there I just spent half an hour [figuring out where this unpacking happens](https://github.com/numpy/numpy/pull/13092#issuecomment-470552955) - if only I had checked all my e-mail! For ufuncs, `__array_wrap__ = None` will prevent the decay. Possibly, this can be generalized? 

Though I agree with @seberg that we ideally just get rid of this behaviour (I tried for just the ufuncs, and lots of failures, but mostly because of `assert_equal` with scalars).

I vaguely recall in an earlier discussion we were meant to have a tracking issue/wiki page or some such to collect pros and cons, difficulties, etc. This seems the way forward...

p.s.  Note that subclasses do not decay to scalars. Maybe somehow one can select a decaying or non-decaying `ndarray` class one wants to use generally.

**Comment by eric-wieser** [comment]
@mhvk, @seberg: Created a new issue at #13105 with my proposal, rather than having this discussion in a place that no-one will ever find again.

**Comment by jakirkham** [comment]
@eric-wieser, I'm afraid I'm not understanding [your comment above]( https://github.com/numpy/numpy/issues/13100#issuecomment-470412475 ). Can you please help clarify it for me? Here's what I'm seeing. How would this example change to isolate the difference between these two?

```python
In [1]: import numpy as np

In [2]: a = np.array(0.5)

In [3]: np.floor(a)
Out[3]: 0.0

In [4]: np.fix(a)
Out[4]: 0.0
```

**Comment by eric-wieser** [comment]
The point I'm trying to make above is that a design flaw in `ufunc.__call__` is forcing us to call `asanyarray` which dooms dask. My proposed alternative does not call this, so would work out of the box.

**Comment by jakirkham** [comment]
Thanks for clarifying. Agree that calling `asanyarray` is not good for compatibility with other array-like libraries.

Though what I'm trying to understand is whether `np.fix` has behavior that deviates from the typical `ufunc` such that it would be tricky to implement as a `ufunc`. Right now I'm not seeing that, but I may be missing something.

**Comment by eric-wieser** [comment]
No, there is nothing special about fix that makes it hard to make a ufunc.

**Comment by jakirkham** [comment]
Thanks for the clarification. Given that it is tricky to get the Python implementation right and having a `ufunc` for `fix` would address the issue, would a PR to convert `fix` to a `ufunc` be accepted?

**Comment by jakirkham** [comment]
Thoughts on the question above?

**Comment by jakirkham** [comment]
Friendly nudge 😉

**Comment by mattip** [comment]
FWIW, I am -1 on a new ufunc. `fix` now has a  `__array_function__` wrapper, can that solve the problem for Dask/CuPy?

**Comment by mhvk** [comment]
I think with `__array_function__`, one can indeed point to another way to override the function, which means the question becomes more purely whether it is worth writing the ufunc. 

My own sense is that it may well be -- at least, I find it somewhat annoying that types of functions that conceptually are universal functions (in acting on single elements) behave slightly differently (e.g., in not having `where` or a way to specify `dtype`; and need a different override) -- but it might be better to do a whole set of related functions in one go, so one can be sure that the C code is reasonably organized. And if one can rewrite the python code such that it only uses ufuncs, that would be fine too.

**Comment by rgommers** [comment]
Quoting the first reply by @eric-wieser:

> It's not clear to me that converting every function to a ufunc is a good idea - they're harder to maintain than to the python functions. 

I fully agree with this. It particularly applies to functions like `np.fix`, that are not used much.

Also, this is now code churn for very little (if any) gain, given that the original problem is solved via `__array_function__`.

**Comment by mhvk** [comment]
After looking again at @eric-wieser's suggestion in #13105 for avoiding the cast to scalar, I think it would be quite do-able to have most functions just written purely in terms of ufuncs, in which case the issue disappears even without `__array_function__`.

**Comment by eric-wieser** [comment]
I think we need to focus on trying to make it easy to write simple ufunc compositions in pure-python, rather than converting all of our code to ufuncs before we get a chance to try. `fix` looks like a great example to address with #13105 or alternative approaches to the problem, so I'd rather we not make it disappear any time soon.

**Comment by mhvk** [comment]
Agreed with that, at least for things like `np.fix` (I'm still playing around with code to chain ufuncs programmatically; it would be a good case for that too)
